shell/common/crash_keys.cc

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2020 Slack Technologies, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/common/crash_keys.h"
+
+#include <deque>
+#include <vector>
+
+#include "base/command_line.h"
+#include "base/environment.h"
+#include "base/no_destructor.h"
+#include "base/strings/string_split.h"
+#include "components/crash/core/common/crash_key.h"
+#include "content/public/common/content_switches.h"
+#include "shell/common/electron_constants.h"
+
+namespace electron {
+
+namespace crash_keys {
+
+namespace {
+
+using ExtraCrashKeys = std::deque<crash_reporter::CrashKeyString<128>>;
+ExtraCrashKeys& GetExtraCrashKeys() {
+  static base::NoDestructor<ExtraCrashKeys> extra_keys;
+  return *extra_keys;
+}
+
+std::deque<std::string>& GetExtraCrashKeyNames() {
+  static base::NoDestructor<std::deque<std::string>> crash_key_names;
+  return *crash_key_names;
+}
+
+}  // namespace
+
+void SetCrashKey(const std::string& key, const std::string& value) {
+  auto& crash_key_names = GetExtraCrashKeyNames();
+
+  auto iter = std::find(crash_key_names.begin(), crash_key_names.end(), key);
+  if (iter == crash_key_names.end()) {
+    crash_key_names.emplace_back(key);
+    GetExtraCrashKeys().emplace_back(crash_key_names.back().c_str());
+    iter = crash_key_names.end() - 1;
```

Comment:
Should this break here instead ?

Reply:
No, this is setting the _key_, and the bit at the end of the function sets the _value_.

This function is a bit weird because it's ensuring that all the crash key names + values are stored in memory that will never be freed, as required by the crash key system.